### ISSUE TITLE
--strictNullChecks issue with inheritance

### ISSUE THREAD

**Issue by yahiko00** [Microsoft/TypeScript#10150]
**TypeScript Version:**  2.0.0

With a basic inheritance issue, it seems that the `--strictNullChecks` flag reports misleading error messages.

**Code**
Let us assume we have a simplified declaration file (from [Phaser game engine](https://github.com/photonstorm/phaser)) as the following:

``` ts
// pixi.d.ts

declare namespace PIXI {
    export class Point {
    }

    export class Matrix {
        append(matrix: Matrix): Matrix;
        apply(pos: Point, newPos: Point): Point;
        applyInverse(pos: Point, newPos: Point): Point;
    }
}
```

It defines two classes, simplified to the minimum for this issue report.
Let us assume we have another declaration file like the following which extends the two previous classes

``` ts
// phaser.d.ts

declare namespace Phaser {
    class Point extends PIXI.Point {
        type: number;
    }

    class Matrix extends PIXI.Matrix {
        type: number;

        append(matrix: Phaser.Matrix): Phaser.Matrix;
        apply(pos: Phaser.Point, newPos?: Phaser.Point): Phaser.Point;
        applyInverse(pos: Phaser.Point, newPos?: Phaser.Point): Phaser.Point;
    }
}
```

**Expected behavior:**
I would expect in such a case the compiler complains there are incompatible types in the signature of methods `apply()` and `applyInverse()` because of optional parameters in the child class whereas they are not in the parent class.

Something like that:

```
phaser.d.ts(8,11): error TS2415: Class 'Phaser.Matrix' incorrectly extends base class 'PIXI.Matrix'.
  Types of property 'apply' are incompatible.
    Type '(pos: Point, newPos?: Point | undefined) => Point' is not assignable to type '(pos: Point, newPos: Point) => Point'.
      Types of parameters 'newPos' and 'newPos' are incompatible.
        Type 'Point' is not assignable to type 'Point | undefined'.
          Type 'PIXI.Point' is not assignable to type 'Phaser.Point'.
            Property 'type' is missing in type 'Point'.
```

**Actual behavior:** 
Instead, the compiler displays misleading error message about the `append()` method which is perfectly correct.

```
../Phaser/typescript/phaser.d.ts(2529,11): error TS2415: Class 'Phaser.Matrix' incorrectly extends base class 'PIXI.Matrix'.
  Types of property 'append' are incompatible.
    Type '(matrix: Matrix) => Matrix' is not assignable to type '(matrix: Matrix) => Matrix'.
      Types of parameters 'matrix' and 'matrix' are incompatible.
        Type 'PIXI.Matrix' is not assignable to type 'Phaser.Matrix'.
          Property 'type' is missing in type 'Matrix'.
```

To get the right error message, I have to put in comments the method `append()`, or the attribut `type` in the Phaser.Matrix class.

**Comment by yahiko00** [comment]
photonstorm/phaser#2673

**Comment by sandersn** [comment]
When I paste those two snippets into the two files in a directory with `--strictNullChecks` I can't repro the issue. Can you check the repro and see if you cut some crucial detail by mistake?

**Comment by sandersn** [comment]
Never mind, I got it. I made `newPos` optional in `PIXI.Matrix` by mistake.

**Comment by yahiko00** [comment]
This issue seems a little bit more than a simple non relevant error message.

I assumed in my original post that there was an error but the error message was misleading. In fact, it seems that there should not be an error at all.

For instance:
```ts
class A {
	method(x: number, y: number): void {
	}
}

class B extends A {
	method(x: number, y?: number): void {
		if (y === undefined) y = 0;
		super.method(x, y);
	}
}
```

In the code above, class `B` overloads `method` turning parameter `y` optional.
Even with the `--strictNullChecks` flag, it works whereas the original example still does not, even with version 2.0.2.

**Comment by Majirefy** [comment]
Any progress on this?

**Comment by sandersn** [comment]
A few notes:
This is three separate issues:

1. The second error line should mention `apply` not `append`.
2. The subsequent error lines should point out that property `type` is missing in `PIXI.Point` not `Matrix`. It's OK for a subclass to have additional properties, so this is a red herring.
3. There shouldn't be an error at all &mdash; the subclass assignability check allows methods to take parameters that are subtypes of the base method's parameter types. So even the 'correct' error obtained by switching the order of `apply` and `append` should not be issued.

(1) is due to depth-first error reporting of assignability checks. That architecture is unlikely to change. The (ugly) workaround is to reorder the methods.

I'm not sure what causes (2) yet. I suspect that the actual error is different &mdash; see (3) &mdash; and that error reporting incorrectly reports that `Phaser.Matrix` is not *assignable* to `PIXI.Matrix`, even though it was supposed to be checking *subclass-assignability*.

(3) is the worst bug since we wouldn't even be reporting errors if it didn't exist. Here's a simpler repro with literal types. `12` is a subtype of `number`, so this is equivalent to `Phaser.Point` and `PIXI.Point`, but the error elaboration is simpler:

```ts
declare class Base {
    x: number;
    m(o: Base): void;
}
declare class Derived extends Base {
    x: 12;
    m(o: Derived | undefined): void;
}
```

Note that `o: Derived | undefined` is essentially equivalent to `o?: Derived` with `--strictNullChecks`. It's just more explicit. `o?: Derived` is still better for real use, though, because it behaves the same way without `--strict`, and it's shorter.

I need to re-read the code to confirm, but I'm pretty sure that the subclass check is a more permissive version of the assignability check. The difference is that `m(o: Derived)` is allowed in `Derived` since it's an override even though `Base` isn't actually assignable to `Derived` because `x: number` is not assignable to `x: 12`.

It looks like the `| undefined` triggers the check to start using the normal assignability checking instead of subclass assignability. This would explain (2) as well &mdash; the error is probably wrong because it's the first error you find when checking for *normal* assignability errors, not subclass assignability errors.

I'll go confirm that this is the case and see whether I can fix it.

**Comment by sandersn** [comment]
Well, after some debugging I discovered that this error is (unfortunately) correct. Surprising, but correct. So this is not an implementation bug and it's debatable whether it's a spec bug. It derives from the way assignability and unions behave, both of which are reasonable in isolation.

Here's an even smaller example that shows why:

```ts
declare class Base {
  x: number
}
declare class Derived extends Base {
  x: 12
}
let a: Base
let b: Derived | string
```

Given these declarations, `b = a` is not allowed because you can't assign a superclass (Base) to a subclass variable (Derived). Normally, `a = b` is allowed because `Derived`, the subclass, is assignable to `Base`, the superclass. However, since we declared `b: Derived | string`, we need to assign both `Derived` *and* `string` to `Base`. And `string` isn't assignable to `Base`. The same holds for `Derived | undefined`, it's just easier for me to think about `Derived | string`.

As I said earlier, subclass-assignability is similar to normal assignability except that it is bivariant. Basically that means that it checks both `Base = Derived` and `Derived = Base` and if *either* is true then it says that `Derived` is a valid subclass. Note that `Derived = Base` violates the Liskov Substitution Principle, so to be correct, we should allow only `Base = Derived` checks.

Unfortunately, adding `undefined` to `m`'s parameter to get `o: Derived | undefined` breaks both `Base = Derived | undefined` and `Derived | undefined = Base` for the same reason that neither `a = b` or `b = a` are legal; `Base` is not assignable to `Derived | undefined` because `x` might not be 12. `Derived | undefined` is not assignable to `Base` because it might be `undefined`, which is not assignable to `Base`.

**Comment by yahiko00** [comment]
Thanks (a little bit late) @sandersn  for all your investigations. Not sure I've understood everything though... ^^
